### PR TITLE
server: support causal-LM rerankers via logit-margin scoring

### LINKED ISSUE #25447: Eval bug: /v1/rerank returns near-zero scores for Qwen3-based rerankers
### Name and Version

❯ .\build\bin\llama-server.exe --version                                     
version: 9924 (90e0f5cfc)
built with GNU 15.2.0 for Windows AMD64

### Operating systems

Windows

### GGML backends

CPU

### Hardware

AMD Ryzen AI 7 Pro 350, 64GB DDR5

### Models

zerank-2 Q6_K:  https://huggingface.co/mradermacher/zerank-2-GGUF
(Qwen3ForCausalLM architecture)

### Problem description & steps to reproduce

Loading a Qwen3-based reranker (zerank-2) with `--reranking` and sending requests to `/v1/rerank` returns scores like `1e-20` for every document. All scores are effectively zero regardless of relevance.

The model works fine through `/v1/completions` with the Qwen3-Reranker judge prompt it correctly scores relevant documents high and irrelevant ones low. So the issue is with the `/v1/rerank` endpoint, not the model.

I think the problem is that the rerank endpoint uses pooling to get a score from `cls`/`cls_out` tensors, but Qwen3 rerankers don't have those as they score by generating "yes" or "no" tokens instead.

Reproduce:
```bash
llama-server -m zerank-2.Q6_K.gguf --reranking --port 18080

curl -s -X POST http://localhost:18080/v1/rerank \
  -H "Content-Type: application/json" \
  -d '{
    "query": "Machine learning is",
    "documents": [
      "Machine learning is a field of study in artificial intelligence.",
      "Paris is the capital of France."
    ]
  }'
```

Result: both documents get near-zero scores:

```json
{
  "results": [
    { "index": 0, "relevance_score": 1.22e-20 },
    { "index": 1, "relevance_score": 2.45e-23 }
  ]
}
```

Same model through `/v1/completions` with the judge prompt gives 0.98 for the ML doc and 0.01 for Paris.

### First Bad Commit

Not a regression, causal-LM rerankers were never supported by this endpoint.

### Relevant log output

<details>
<summary>Logs</summary>

```console
llama_init_from_model: model default pooling_type is [-1], but [4] was specified
srv    load_model: initializing, n_slots = 4, n_ctx_slot = 40960
srv  llama_server: model loaded
```

</details>

### PR DESCRIPTION
**Pull request by huveewomg**
## Overview

Fixes #25447

The `/v1/rerank` endpoint silently returns near-zero scores (~1e-20) for causal-LM rerankers like zerank-2 and Qwen3-Reranker. These models have no classification head, their `lm_head` is tied to token embeddings therefore the pooling path extracts raw hidden states instead of classification scores.

This PR adds a completion-based scoring path for such models. When `llama_model_has_cls_head()` returns false, the endpoint generates one token per document using a judge prompt and computes relevance from the yes/no logit margin. Existing BERT-style rerankers (jina, bge) are unaffected, they have `cls`/`cls_out` tensors and continue using the pooling path.

Changes:

- **`include/llama.h` / `src/llama-model.cpp`**: add `llama_model_has_cls_head()` to check for `cls` or `cls_out` tensors
- **`tools/server/server-context.cpp`**: detect causal-LM rerankers and route through completion-based logit-margin scoring. Uses the model's GGUF "rerank" chat template when available, falls back to a built-in Qwen3-Reranker judge prompt
- **`tools/server/tests/unit/test_rerank_causal.py`**:  test cases for causal-LM reranking + pooling path regression test

Tested with zerank-2 Q6_K (before: 1e-20 garbage, after: 0.984/0.016 matching `/v1/completions`) and jina-reranker-v1-tiny-en F16 (pooling path unchanged, correct scores).

## Reproduction & results

**Before (pooling path returns garbage for causal-LM reranker):**

```bash
llama-server -m zerank-2.Q6_K.gguf --reranking --port 18080

curl -s -X POST http://localhost:18080/v1/rerank \
  -H "Content-Type: application/json" \
  -d '{"query":"Machine learning is","documents":["Machine learning is a field of study in artificial intelligence.","Paris is the capital of France."]}'

{
  "results": [
    { "index": 0, "relevance_score": 1.22e-20 },
    { "index": 1, "relevance_score": 2.45e-23 }
  ]
}

After (logit-margin scoring, same model, same endpoint):

{
  "results": [
    { "index": 0, "relevance_score": 0.9844 },
    { "index": 1, "relevance_score": 0.0156 }
  ]
}

jina-reranker-v1-tiny-en (BERT-style, pooling path unchanged):

{
  "results": [
    { "index": 0, "relevance_score": 0.1004 },
    { "index": 1, "relevance_score": -0.0026 }
  ]
}
```

## Requirements

- I have read and agree with the [contributing guidelines](https://github.com/ggml-org/llama.cpp/blob/master/CONTRIBUTING.md)
- AI usage disclosure: YES, used for research assistance. Code manually reviewed and tested on both model types.

### CODE REVIEW

**Comment by ngxson** [comment]
thanks for the contribution, but hard-coding a synthetic prompt in llama-server is a no-go

this should better be done in downstream code using the provided output logits (logprob), closing it now as an out-of-scope feature
